### ISSUE TITLE
ARM64: several test case failure in runtime-tests.sh

### ISSUE THREAD

**Issue by lostjeffle** [iovisor/bpftrace#1137]
OS: 4.19.48
arch: aarch64
bpftrace: v0.9.3

I got several test case failure when testing the bpftrace rpm package, which is built by myself.

All failed test cases include:
- array.array_element_access_-_assign_to_map
- array.array_element_access_-_assign_to_var
- builtin.sarg
- call.str
- intptrcast.Casting_ints
- intptrcast.Sum_casted_value
- json-output.join_delim
- regression.logical_and_or_different_sizes

For example, the error log of "regression.logical_and_or_different_sizes" is:

```
[ RUN      ] regression.logical_and_or_different_sizes
[  FAILED  ] regression.logical_and_or_different_sizes
	Command: /usr/bin/bpftrace -v -e 'struct Foo { int m; } uprobe:./testprogs/simple_struct:func { $foo = (struct Foo*)arg0; printf("%d %d %d %d %d¥n", $foo->m, $foo->m && 0, 1 && $foo->m, $foo->m || 0, 0 || $foo->m); exit(); }'
	Expected: 2 0 1 1 1
	Found: Attaching 1 probe...

Program ID: 1103

Bytecode: 
0: (bf) r6 = r1
1: (79) r7 = *(u64 *)(r6 +0)
2: (b7) r8 = 0
3: (7b) *(u64 *)(r10 -48) = r8
4: (7b) *(u64 *)(r10 -56) = r8
5: (bf) r1 = r10
6: (07) r1 += -4
7: (b7) r2 = 4
8: (bf) r3 = r7
9: (85) call bpf_probe_read#4
10: (61) r1 = *(u32 *)(r10 -4)
11: (63) *(u32 *)(r10 -48) = r1
12: (bf) r1 = r10
13: (07) r1 += -4
14: (b7) r2 = 4
15: (bf) r3 = r7
16: (85) call bpf_probe_read#4
17: (7b) *(u64 *)(r10 -40) = r8
18: (bf) r1 = r10
19: (07) r1 += -4
20: (b7) r2 = 4
21: (bf) r3 = r7
22: (85) call bpf_probe_read#4
23: (b7) r8 = 1
24: (61) r2 = *(u32 *)(r10 -4)
25: (b7) r1 = 1
26: (55) if r2 != 0x0 goto pc+1
 R0=inv(id=0) R1=inv1 R2=inv0 R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-40=0 fp-48=0 fp-56=0
27: (b7) r1 = 0
28: (7b) *(u64 *)(r10 -32) = r1
29: (bf) r1 = r10
30: (07) r1 += -4
31: (b7) r2 = 4
32: (bf) r3 = r7
33: (85) call bpf_probe_read#4
34: (61) r2 = *(u32 *)(r10 -4)
35: (b7) r1 = 1
36: (55) if r2 != 0x0 goto pc+1
 R0=inv(id=0) R1=inv1 R2=inv0 R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-32=0 fp-40=0 fp-48=0 fp-56=0
37: (b7) r1 = 0
38: (7b) *(u64 *)(r10 -24) = r1
39: (bf) r1 = r10
40: (07) r1 += -4
41: (b7) r2 = 4
42: (bf) r3 = r7
43: (85) call bpf_probe_read#4
44: (61) r1 = *(u32 *)(r10 -4)
45: (55) if r1 != 0x0 goto pc+1
 R0=inv(id=0) R1=inv0 R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-24=0 fp-32=0 fp-40=0 fp-48=0 fp-56=0
46: (b7) r8 = 0
47: (7b) *(u64 *)(r10 -16) = r8
48: (18) r7 = 0xffffa05ffb177000
50: (85) call bpf_get_smp_processor_id#8
51: (bf) r4 = r10
52: (07) r4 += -56
53: (bf) r1 = r6
54: (bf) r2 = r7
55: (bf) r3 = r0
56: (b7) r5 = 48
57: (85) call bpf_perf_event_output#25
58: (b7) r1 = 30000
59: (7b) *(u64 *)(r10 -56) = r1
60: (18) r7 = 0xffffa05ffb177000
62: (85) call bpf_get_smp_processor_id#8
63: (bf) r4 = r10
64: (07) r4 += -56
65: (bf) r1 = r6
66: (bf) r2 = r7
67: (bf) r3 = r0
68: (b7) r5 = 8
69: (85) call bpf_perf_event_output#25
70: (b7) r0 = 0
71: (95) exit

from 45 to 47: R0=inv(id=0) R1=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-24=0 fp-32=0 fp-40=0 fp-48=0 fp-56=0
47: (7b) *(u64 *)(r10 -16) = r8
48: (18) r7 = 0xffffa05ffb177000
50: (85) call bpf_get_smp_processor_id#8
51: (bf) r4 = r10
52: (07) r4 += -56
53: (bf) r1 = r6
54: (bf) r2 = r7
55: (bf) r3 = r0
56: (b7) r5 = 48
57: (85) call bpf_perf_event_output#25
58: safe

from 36 to 38: R0=inv(id=0) R1=inv1 R2=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-32=0 fp-40=0 fp-48=0 fp-56=0
38: (7b) *(u64 *)(r10 -24) = r1
39: (bf) r1 = r10
40: (07) r1 += -4
41: (b7) r2 = 4
42: (bf) r3 = r7
43: (85) call bpf_probe_read#4
44: (61) r1 = *(u32 *)(r10 -4)
45: (55) if r1 != 0x0 goto pc+1
 R0=inv(id=0) R1=inv0 R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-32=0 fp-40=0 fp-48=0 fp-56=0
46: (b7) r8 = 0
47: (7b) *(u64 *)(r10 -16) = r8
48: (18) r7 = 0xffffa05ffb177000
50: (85) call bpf_get_smp_processor_id#8
51: (bf) r4 = r10
52: (07) r4 += -56
53: (bf) r1 = r6
54: (bf) r2 = r7
55: (bf) r3 = r0
56: (b7) r5 = 48
57: (85) call bpf_perf_event_output#25
58: safe

from 45 to 47: R0=inv(id=0) R1=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-32=0 fp-40=0 fp-48=0 fp-56=0
47: (7b) *(u64 *)(r10 -16) = r8
48: (18) r7 = 0xffffa05ffb177000
50: (85) call bpf_get_smp_processor_id#8
51: (bf) r4 = r10
52: (07) r4 += -56
53: (bf) r1 = r6
54: (bf) r2 = r7
55: (bf) r3 = r0
56: (b7) r5 = 48
57: (85) call bpf_perf_event_output#25
58: safe

from 26 to 28: R0=inv(id=0) R1=inv1 R2=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-40=0 fp-48=0 fp-56=0
28: (7b) *(u64 *)(r10 -32) = r1
29: (bf) r1 = r10
30: (07) r1 += -4
31: (b7) r2 = 4
32: (bf) r3 = r7
33: (85) call bpf_probe_read#4
34: (61) r2 = *(u32 *)(r10 -4)
35: (b7) r1 = 1
36: (55) if r2 != 0x0 goto pc+1
 R0=inv(id=0) R1=inv1 R2=inv0 R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-40=0 fp-48=0 fp-56=0
37: (b7) r1 = 0
38: (7b) *(u64 *)(r10 -24) = r1
39: (bf) r1 = r10
40: (07) r1 += -4
41: (b7) r2 = 4
42: (bf) r3 = r7
43: (85) call bpf_probe_read#4
44: (61) r1 = *(u32 *)(r10 -4)
45: (55) if r1 != 0x0 goto pc+1
 R0=inv(id=0) R1=inv0 R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-24=0 fp-40=0 fp-48=0 fp-56=0
46: (b7) r8 = 0
47: (7b) *(u64 *)(r10 -16) = r8
48: (18) r7 = 0xffffa05ffb177000
50: (85) call bpf_get_smp_processor_id#8
51: (bf) r4 = r10
52: (07) r4 += -56
53: (bf) r1 = r6
54: (bf) r2 = r7
55: (bf) r3 = r0
56: (b7) r5 = 48
57: (85) call bpf_perf_event_output#25
58: safe

from 45 to 47: R0=inv(id=0) R1=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-24=0 fp-40=0 fp-48=0 fp-56=0
47: (7b) *(u64 *)(r10 -16) = r8
48: (18) r7 = 0xffffa05ffb177000
50: (85) call bpf_get_smp_processor_id#8
51: (bf) r4 = r10
52: (07) r4 += -56
53: (bf) r1 = r6
54: (bf) r2 = r7
55: (bf) r3 = r0
56: (b7) r5 = 48
57: (85) call bpf_perf_event_output#25
58: safe

from 36 to 38: R0=inv(id=0) R1=inv1 R2=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-40=0 fp-48=0 fp-56=0
38: (7b) *(u64 *)(r10 -24) = r1
39: (bf) r1 = r10
40: (07) r1 += -4
41: (b7) r2 = 4
42: (bf) r3 = r7
43: (85) call bpf_probe_read#4
44: (61) r1 = *(u32 *)(r10 -4)
45: (55) if r1 != 0x0 goto pc+1
 R0=inv(id=0) R1=inv0 R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-40=0 fp-48=0 fp-56=0
46: (b7) r8 = 0
47: (7b) *(u64 *)(r10 -16) = r8
48: (18) r7 = 0xffffa05ffb177000
50: (85) call bpf_get_smp_processor_id#8
51: (bf) r4 = r10
52: (07) r4 += -56
53: (bf) r1 = r6
54: (bf) r2 = r7
55: (bf) r3 = r0
56: (b7) r5 = 48
57: (85) call bpf_perf_event_output#25
58: safe

from 45 to 47: R0=inv(id=0) R1=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv1 R10=fp0,call_-1 fp-40=0 fp-48=0 fp-56=0
47: (7b) *(u64 *)(r10 -16) = r8
48: (18) r7 = 0xffffa05ffb177000
50: (85) call bpf_get_smp_processor_id#8
51: (bf) r4 = r10
52: (07) r4 += -56
53: (bf) r1 = r6
54: (bf) r2 = r7
55: (bf) r3 = r0
56: (b7) r5 = 48
57: (85) call bpf_perf_event_output#25
58: safe
processed 184 insns (limit 131072), stack depth 56

Attaching uprobe:./testprogs/simple_struct:func
Running...
0 0 0 0 0
```

It seems that there may be a problem with register parsing. I'm not sure if it is related with my testing environment or compiling options. The same package (build from v0.9.3, but for x86_64) passed all test cases in x86_64.

The complete log is attached, and any suggestions would be appreciated.

**Comment by lostjeffle** [comment]
The complete log is attached.
[bpftrace-test.txt](https://github.com/iovisor/bpftrace/files/4179277/bpftrace-test.txt)

**Comment by mmisono** [comment]
I don't check all, but at least [intptrcast test](https://github.com/iovisor/bpftrace/blob/master/tests/runtime/intptrcast) relies on x86_64 ABI (it tries to access an argument on the stack). Is there any way to run a runtime test only on a specific architecture?

**Comment by lostjeffle** [comment]
> I don't check all, but at least [intptrcast test](https://github.com/iovisor/bpftrace/blob/master/tests/runtime/intptrcast) relies on x86_64 ABI (it tries to access an argument on the stack). Is there any way to run a runtime test only on a specific architecture?

Yes, 'intptrcast' should be compatible only with x86_64 ABI rather than aarch64. But the 'regression.logical_and_or_different_sizes' I mentioned above should not be architecture specific. A little confused.

**Comment by mmisono** [comment]
> 1: (79) r7 = *(u64 *)(r6 +0)

This code assumes that the first argument of the function is in the x0 register (to be precise, the first field of the `struct pt_regs`). Is it true for your architecture? (I guess so, but just to be sure)
It can be checked by disassembling ./testprogs/simple_struct, which is located in the build directory.

**Comment by lostjeffle** [comment]
> > 1: (79) r7 = *(u64 *)(r6 +0)
> 
> This code assumes that the first argument of the function is in the x0 register (to be precise, the first field of the `struct pt_regs`). Is it true for your architecture? (I guess so, but just to be sure)
> It can be checked by disassembling ./testprogs/simple_struct, which is located in the build directory.

Yes x0 serves as the register containing the first argument of the function. The disassembling of simple_struct is 

```
000000000040086c <func@@Base>:
  40086c:       d10043ff        sub     sp, sp, #0x10
  400870:       f90007e0        str     x0, [sp, #8]
  400874:       f94007e0        ldr     x0, [sp, #8]
  400878:       b9400000        ldr     w0, [x0]
  40087c:       910043ff        add     sp, sp, #0x10
  400880:       d65f03c0        ret

0000000000400884 <main@@Base>:
  400884:       a9be7bfd        stp     x29, x30, [sp, #-32]!
  400888:       910003fd        mov     x29, sp
  40088c:       52800040        mov     w0, #0x2                        // #2
  400890:       b9001be0        str     w0, [sp, #24]
  400894:       910063e0        add     x0, sp, #0x18
  400898:       97fffff5        bl      40086c <func@@Base>
  40089c:       52800000        mov     w0, #0x0                        // #0
  4008a0:       a8c27bfd        ldp     x29, x30, [sp], #32
  4008a4:       d65f03c0        ret
```

**Comment by mmisono** [comment]
Hm... the generated BPF program is almost the same as what generated on x86_64 except for the argument access (what I mentioned above).

Did you try to run the test program manually? It can be run in the build directory by:

```sh
% sudo ./src/bpftrace -v -e 'struct Foo { int m; } uprobe:./tests/testprogs/simple_struct:func { $foo = (struct Foo*)arg0; printf("%d %d %d %d %d\n", $foo->m, $foo->m
&& 0, 1 && $foo->m, $foo->m || 0, 0 || $foo->m); exit(); }'
```

then run `simple_struct` in another teminal.
```
% ./tests/testprogs/simple_struct
```

**Comment by lostjeffle** [comment]
@mmisono I tried manually but got the same error.

```
Running...
0 0 0 0 0
```

**Comment by mmisono** [comment]
I'm not sure what is happening...  Could you try to run the following bcc program to check the return value of `bpf_probe_read()`? (when `bpf_probe_read()` failed, bpftrace treats a value as zero).

```python
#!/usr/bin/env python

import bcc

text = r"""
#include <uapi/linux/ptrace.h>

struct Foo { int m; };

int func(struct pt_regs *ctx) {
  struct Foo *foo = (struct Foo*)PT_REGS_PARM1(ctx);
  if (!foo) {
    bpf_trace_printk("foo is null\n");
    return 0;
  }
  int m;
  int r = bpf_probe_read(&m, sizeof(m), foo);
  if (r != 0)
    bpf_trace_printk("bpf_probe_read failed\n");
  else
    bpf_trace_printk("foo->m = %d\n", m);
  return 0;
}
"""


b = bcc.BPF(text=text, debug=0x8)
b.attach_uprobe(name="./tests/testprogs/simple_struct",
                sym="func", fn_name="func")
b.trace_print()
```

and then run `simple_struct` in another teminal.

**Comment by fbs** [comment]
Might be related to #614

**Comment by lostjeffle** [comment]
It seems 'bpf_probe_read failed'.
@mmisono 

```
Disassembly of section .bpf.fn.func:
func:
; int func(struct pt_regs *ctx) { // Line  17
   0:	79 13 00 00 00 00 00 00 	r3 = *(u64 *)(r1 + 0)
; if (!foo) { // Line  20
   1:	55 03 0c 00 00 00 00 00 	if r3 != 0 goto +12
   2:	b7 01 00 00 00 00 00 00 	r1 = 0
; ({ char _fmt[] = "foo is null\n"; bpf_trace_printk_(_fmt, sizeof(_fmt)); }); // Line  21
   3:	73 1a ec ff 00 00 00 00 	*(u8 *)(r10 - 20) = r1
   4:	b7 01 00 00 75 6c 6c 0a 	r1 = 174877813
   5:	63 1a e8 ff 00 00 00 00 	*(u32 *)(r10 - 24) = r1
   6:	18 01 00 00 66 6f 6f 20 00 00 00 00 69 73 20 6e 	r1 = 7935469438779748198 ll
   8:	7b 1a e0 ff 00 00 00 00 	*(u64 *)(r10 - 32) = r1
   9:	bf a1 00 00 00 00 00 00 	r1 = r10
  10:	07 01 00 00 e0 ff ff ff 	r1 += -32
  11:	b7 02 00 00 0d 00 00 00 	r2 = 13
  12:	85 00 00 00 06 00 00 00 	call 6
  13:	05 00 24 00 00 00 00 00 	goto +36
  14:	bf a1 00 00 00 00 00 00 	r1 = r10
; int r = bpf_probe_read(&m, sizeof(m), foo); // Line  25
  15:	07 01 00 00 fc ff ff ff 	r1 += -4
  16:	b7 02 00 00 04 00 00 00 	r2 = 4
  17:	85 00 00 00 04 00 00 00 	call 4
  18:	67 00 00 00 20 00 00 00 	r0 <<= 32
  19:	77 00 00 00 20 00 00 00 	r0 >>= 32
; if (r != 0) // Line  26
  20:	15 00 11 00 00 00 00 00 	if r0 == 0 goto +17
  21:	b7 01 00 00 00 00 00 00 	r1 = 0
; ({ char _fmt[] = "bpf_probe_read failed\n"; bpf_trace_printk_(_fmt, sizeof(_fmt)); }); // Line  27
  22:	73 1a f6 ff 00 00 00 00 	*(u8 *)(r10 - 10) = r1
  23:	b7 01 00 00 64 0a 00 00 	r1 = 2660
  24:	6b 1a f4 ff 00 00 00 00 	*(u16 *)(r10 - 12) = r1
  25:	b7 01 00 00 61 69 6c 65 	r1 = 1701603681
  26:	63 1a f0 ff 00 00 00 00 	*(u32 *)(r10 - 16) = r1
  27:	18 01 00 00 65 5f 72 65 00 00 00 00 61 64 20 66 	r1 = 7358992160599990117 ll
  29:	7b 1a e8 ff 00 00 00 00 	*(u64 *)(r10 - 24) = r1
  30:	18 01 00 00 62 70 66 5f 00 00 00 00 70 72 6f 62 	r1 = 7093013765094273122 ll
  32:	7b 1a e0 ff 00 00 00 00 	*(u64 *)(r10 - 32) = r1
  33:	bf a1 00 00 00 00 00 00 	r1 = r10
  34:	07 01 00 00 e0 ff ff ff 	r1 += -32
  35:	b7 02 00 00 17 00 00 00 	r2 = 23
  36:	85 00 00 00 06 00 00 00 	call 6
  37:	05 00 0c 00 00 00 00 00 	goto +12
  38:	b7 01 00 00 00 00 00 00 	r1 = 0
; ({ char _fmt[] = "foo->m = %d\n"; bpf_trace_printk_(_fmt, sizeof(_fmt), m); }); // Line  29
  39:	73 1a ec ff 00 00 00 00 	*(u8 *)(r10 - 20) = r1
  40:	b7 01 00 00 20 25 64 0a 	r1 = 174335264
  41:	63 1a e8 ff 00 00 00 00 	*(u32 *)(r10 - 24) = r1
  42:	18 01 00 00 66 6f 6f 2d 00 00 00 00 3e 6d 20 3d 	r1 = 4404640549386022758 ll
  44:	7b 1a e0 ff 00 00 00 00 	*(u64 *)(r10 - 32) = r1
  45:	61 a3 fc ff 00 00 00 00 	r3 = *(u32 *)(r10 - 4)
  46:	bf a1 00 00 00 00 00 00 	r1 = r10
  47:	07 01 00 00 e0 ff ff ff 	r1 += -32
  48:	b7 02 00 00 0d 00 00 00 	r2 = 13
  49:	85 00 00 00 06 00 00 00 	call 6
; } // Line  31
  50:	b7 00 00 00 00 00 00 00 	r0 = 0
  51:	95 00 00 00 00 00 00 00 	exit


           <...>-111176 [019] .... 624650.748442: 0: bpf_probe_read failed

```

**Comment by mmisono** [comment]
As fbs said, `bpf_probe_read()` may fail to read user memory on some architectures (but I don't know whether aarch64 is one of these architectures.) Linux kernel 5.5 introduces new helper function `bpf_probe_read_user()` and `bpf_probe_read_kernel()` to solve the problem.  Please see [the commit message](https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/commit/?id=6ae08ae3dea2cfa03dd3665a3c8475c2d429ef47) for the detail. If you can upgrade the kernel, you can try `bpf_probe_read_user()` in the bcc program.

@lostjeffle 
Just to be sure, could you tell me the return value of `bpf_probe_read()` by

```
bpf_trace_printk("bpf_probe_read faield: %d\n", r);
```
?

**Comment by lostjeffle** [comment]
@mmisono 
simple_struct-4290  [018] .... 628470.524320: 0: bpf_probe_read failed -14

**Comment by mmisono** [comment]
I found this bss'c issue: iovisor/bcc#2253
After all, it appears that we need to use `bpf_probe_read_user()` to read user memory on aarch64.

--------------

So, there are two issues about the runtime tests:

1. Some testes are only for x86_64
    - We need to exclude such tests on the non-x86_64 environment
2. bpf_probe_read() does not work on some architectures (#614)

**Comment by dalehamel** [comment]
these failures are familiar to what I encountered when worked on #1095, such as some registers /features being hardcoded for x86

**Comment by lostjeffle** [comment]
> As fbs said, `bpf_probe_read()` may fail to read user memory on some architectures (but I don't know whether aarch64 is one of these architectures.) Linux kernel 5.5 introduces new helper function `bpf_probe_read_user()` and `bpf_probe_read_kernel()` to solve the problem. Please see [the commit message](https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/commit/?id=6ae08ae3dea2cfa03dd3665a3c8475c2d429ef47) for the detail. If you can upgrade the kernel, you can try `bpf_probe_read_user()` in the bcc program.

@mmisono I test kernel 5.5 today and the case passed as it prints "foo->m = 2". So yes it can been fixed by bps_probe_read_{usr/kernel}.

**Comment by lostjeffle** [comment]
Indeed there're two things to work with aarch64
 
> 1. Some testes are only for x86_64
>    
>    * We need to exclude such tests on the non-x86_64 environment

Does pre-processing macros such as 'ARCH_X86_64' / 'ARCH_AARCH64' work?

> 2. bpf_probe_read() does not work on some architectures (#614)

Some architecture (such as aarch64) needs bpf_probe_read_{usr/kernel} instead of bpf_probe_read.

**Comment by lostjeffle** [comment]
Besides several 'codegen' tests fail on aarch64. 

codegen.builtin_arg
codegen.builtin_func
codegen.builtin_func_uprobe
codegen.builtin_func_wild
codegen.builtin_retval
codegen.builtin_sarg
codegen.call_override
codegen.call_reg
codegen.call_signal
codegen.call_str
codegen.call_str_2_expr
codegen.call_str_2_lit
codegen.comparison_extend
codegen.intcast_retval
codegen.intcast_call
codegen.intptrcast_assign_var
codegen.intptrcast_call
codegen.struct_char

**Comment by mmisono** [comment]
@lostjeffle 
> Does pre-processing macros such as 'ARCH_X86_64' / 'ARCH_AARCH64' work?

I think adding REQUIRES check to runtime tests can limit tests so that they run only on x86_64 ([example](https://github.com/iovisor/bpftrace/blob/11814f2981c2fcf4c0177720910786c62474e100/tests/runtime/usdt#L5)). Could you try adding the following line to intptrcast and check the result?
```
REQUIRES uname -m | grep -q x86_64
```

-------

I think the following tests needs to add that REQUIRES (since they rely on x86_64 ABI to access args in the stack):

- codegen.intptrcast_call
- codegen.intptrcast_assign_var
- codegen.builtin_sarg

(in other words, other tests fail due to the bpf_probe_read() problem).

**Comment by lostjeffle** [comment]
> REQUIRES uname -m | grep -q x86_64
This works for runtime tests. 

Besides some arch specific code is hardcoded in .cpp，such as semantic_analyser.cpp.

```
TEST(semantic_analyser, builtin_functions)
{
  ...
  test("kprobe:f { reg(\"ip\") }", 0);
  ...
}
```

Conditional compiling may be needed in this kind of issue?

**Comment by mmisono** [comment]
[src/arch](https://github.com/iovisor/bpftrace/tree/master/src/arch) contains architecture-specific implementations. So we can add a function to get a register name for a test there, though I'm not certain that is the best approach.

**Comment by mmisono** [comment]
Just found #593, which tried to add changes to tests for aarch64 but wasn't merged since it introduced many code path changes.
Disabling these x86_64 specific tests on other architectures (by using REQUIRES for runtime tests and conditional compiling for other parts) eases maintainability and I think that is enough for now.

**Comment by lostjeffle** [comment]
- some runtime-test had been fixed by  commit 48d405133160 ("bpftrace: Add architecture specific testcase support")
- opensnoop/statsnoop in tools-parsing-test had been fixed by commit 330953c11190c23a12f98d374d1b5eb6d7aad1fe ("fix build failed on aarch64")
-  semantic_analyser.builtin_functions, semantic_analyser.call_reg, some codegen test cases have been fixed by workaround  in commit 2aa294b12568978eca1251100e149e830b23c097 ("tests: Add architecture specific testcase support")
